src/mbgl/geometry/feature_index.cpp

```diff
@@ -0,0 +1,230 @@
+#include <mbgl/geometry/feature_index.hpp>
+#include <mbgl/util/math.hpp>
+#include <mbgl/style/style.hpp>
+#include <mbgl/style/style_layer.hpp>
+#include <mbgl/layer/symbol_layer.hpp>
+#include <mbgl/util/get_geometries.hpp>
+#include <mbgl/text/collision_tile.hpp>
+#include <mbgl/util/rapidjson.hpp>
+#include <rapidjson/writer.h>
+
+#include <cassert>
+#include <string>
+
+using namespace mbgl;
+
+FeatureIndex::FeatureIndex() {}
+
+void FeatureIndex::insert(const GeometryCollection& geometries, std::size_t index,
+        const std::string& sourceLayerName, const std::string& bucketName) {
+
+    auto sortIndex = treeBoxes.size();
+
+    for (auto& ring : geometries) {
+
+        float minX = std::numeric_limits<float>::infinity();
+        float minY = std::numeric_limits<float>::infinity();
+        float maxX = -std::numeric_limits<float>::infinity();
+        float maxY = -std::numeric_limits<float>::infinity();
+        for (auto& p : ring) {
+            const float x = p.x;
+            const float y = p.y;
+            minX = util::min(minX, x);
+            minY = util::min(minY, y);
+            maxX = util::max(maxX, x);
+            maxY = util::max(maxY, y);
+        }
+
+        treeBoxes.emplace_back(
+            TreeBox {
+                TreePoint { minX, minY },
+                TreePoint { maxX, maxY }
+            },
+            IndexedSubfeature { index, sourceLayerName, bucketName, sortIndex }
+        );
+    }
+}
+
+void FeatureIndex::loadTree() {
+    tree.insert(treeBoxes.begin(), treeBoxes.end());
```

Comment:
Should this loop be made cancellable/interruptible (or not)?
See the comment here:

https://github.com/mapbox/mapbox-gl-native/issues/2644#issuecomment-210258384